lib/blocks.php

```diff
@@ -171,98 +172,107 @@ function gutenberg_render_block( $block ) {
  * @return string          Updated post content.
  */
 function do_blocks( $content ) {
-	global $post;
+	$blocks = gutenberg_parse_blocks( $content );
+	return _recurse_blocks( $blocks, $blocks );
+}
+add_filter( 'the_content', 'do_blocks', 9 ); // BEFORE do_shortcode().
 
-	$rendered_content      = '';
-	$dynamic_block_pattern = get_dynamic_blocks_regex();
+/**
+ * Helper function for do_blocks(), to recurse through the block tree.
+ *
+ * @since 4.1.0
+ * @access private
+ *
+ * @param array $blocks Array of blocks from parse_blocks()
+ * @return string
+ */
+function _recurse_blocks( $blocks, $all_blocks ) {
+	global $post;
 
 	/*
 	 * Back up global post, to restore after render callback.
 	 * Allows callbacks to run new WP_Query instances without breaking the global post.
 	 */
 	$global_post = $post;
 
-	while ( preg_match( $dynamic_block_pattern, $content, $block_match, PREG_OFFSET_CAPTURE ) ) {
-		$opening_tag     = $block_match[0][0];
-		$offset          = $block_match[0][1];
-		$block_name      = $block_match[1][0];
-		$is_self_closing = isset( $block_match[4] );
-
-		// Reset attributes JSON to prevent scope bleed from last iteration.
-		$block_attributes_json = null;
-		if ( isset( $block_match[3] ) ) {
-			$block_attributes_json = $block_match[3][0];
+	$rendered_content = '';
+	$dynamic_blocks   = get_dynamic_block_names();
+
+	foreach ( $blocks as $block ) {
+		$block = (array) $block;
+		if ( in_array( $block['blockName'], $dynamic_blocks ) ) {
+			// Find registered block type. We can assume it exists since we use the
+			// `get_dynamic_block_names` function as a source for pattern matching.
+			$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+
+			// Replace dynamic block with server-rendered output.
+			$block_content = $block_type->render( (array) $block['attrs'], $block['innerHTML'] );
+		} elseif ( $block['innerBlocks'] ) {
```

Comment:
The way these conditionals are laid, blocks can either be Dynamic Blocks, OR they can have innerBlocks.  It doesn't permit Dynamic Blocks to have innerBlocks.  Which seems bad.

Reply:
this is true, hence the reason this PR exists.

I'd rather not render inner blocks first then pass them up because then that limits what we can do with inner blocks. it should be good enough to give blocks their inner blocks in structural form, let them decide if they want to modify or filter them, then render the document.

this PR is just one step in the process - we have lots to overhaul here and maybe I'm wrong but I thought that @mcsf decided to keep this as-is until post merge and leave dynamic blocks the way they are - as in, yesterday or the day before he decided. let's see what he says